basicapp/managedapps.go

```diff
@@ -0,0 +1,216 @@
+	}
+
+	{
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "waiting for deployed status")
+
+		err = ms.resource.WaitForStatus(ms.chart.Name, "DEPLOYED")
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "chart is deployed")
+	}
+	{
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "checking resources")
+
+		for _, ds := range ms.chartResources.DaemonSets {
+			ms.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("checking daemonset %#q", ds.Name))
+
+			err = ms.checkDaemonSet(ds)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			ms.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("daemonset %#q is correct", ds.Name))
+		}
+
+		for _, d := range ms.chartResources.Deployments {
+			ms.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("checking deployment %#q", d.Name))
+
+			err = ms.checkDeployment(d)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			ms.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment %#q is correct", d.Name))
+		}
+
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "resources are correct")
+	}
+
+	{
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "running release tests")
+
+		err = ms.helmClient.RunReleaseTest(ctx, ms.chart.Name)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		ms.logger.LogCtx(ctx, "level", "debug", "message", "release tests passed")
+	}
+
+	return nil
+}
+
+// checkDaemonSet ensures that key properties of the daemonset are correct.
+func (ms *ManagedApps) checkDaemonSet(expectedDaemonSet DaemonSet) error {
+	ds, err := ms.clients.K8sClient().Apps().DaemonSets(expectedDaemonSet.Namespace).Get(expectedDaemonSet.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return microerror.Maskf(notFoundError, "daemonset %#q", expectedDaemonSet.Name)
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = ms.checkLabels("daemonset labels", expectedDaemonSet.Labels, ds.ObjectMeta.Labels)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = ms.checkLabels("daemonset matchLabels", expectedDaemonSet.MatchLabels, ds.Spec.Selector.MatchLabels)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = ms.checkLabels("daemonset pod labels", expectedDaemonSet.Labels, ds.Spec.Template.ObjectMeta.Labels)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+// checkDeployment ensures that key properties of the deployment are correct.
+func (ms *ManagedApps) checkDeployment(expectedDeployment Deployment) error {
+	ds, err := ms.clients.K8sClient().Apps().Deployments(expectedDeployment.Namespace).Get(expectedDeployment.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return microerror.Maskf(notFoundError, "deployment: %#q", expectedDeployment.Name)
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if int32(expectedDeployment.Replicas) != *ds.Spec.Replicas {
```

Comment:
This is dumb and again my bad :(

Can we remove replicas from the spec?  We can just check that the ready replicas matches the number in the spec. WDYT?

Reply:
I think we can compare `.Spec.Replicas` == `.Status.ReadyReplicas`. What do you think?

Reply:
But I don't like it in general since it would take a while until all pods are counted in `ReadyReplica`.

Reply:
Up to you but I'm fine with waiting. For complex apps like Ingress Controller I think its a better test as the pods should be ready.

Reply:
@rossf7 Added retry mechanism. PTAL.